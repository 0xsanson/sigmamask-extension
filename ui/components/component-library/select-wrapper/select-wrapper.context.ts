import { useContext, createContext } from 'react';
import { SelectContextType } from './select-wrapper.types';

export const SelectContext = createContext<SelectContextType | undefined>(
  undefined,
);

export const useSelectContext = () => {
  const context = useContext(SelectContext);

  if (!context) {
    throw new Error('useSelectContext must be used within a SelectWrapper');
  }

  return context;
};

// Custom hook to access the uncontrolledValue
export function useUncontrolledValue() {
  const context = useContext(SelectContext);
  if (!context) {
    throw new Error('useUncontrolledValue must be used within a SelectWrapper');
  }
  return context.uncontrolledValue;
}
