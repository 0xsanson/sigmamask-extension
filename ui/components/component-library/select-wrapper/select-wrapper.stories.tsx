import React from 'react';
import { StoryFn, Meta } from '@storybook/react';
import { SelectButton } from '../select-button';
import { SelectOption } from '../select-option';
import { Button } from '../button';
import { Text } from '../text';
import { BackgroundColor } from '../../../helpers/constants/design-system';
import README from './README.mdx';

import { SelectWrapper, useSelectContext } from '.';

export default {
  title: 'Components/ComponentLibrary/SelectWrapper',
  component: SelectWrapper,
  parameters: {
    docs: {
      page: README,
      story: {
        inline: false,
      },
    },
  },
  argTypes: {
    className: {
      control: 'text',
    },
    children: {
      control: 'text',
    },
    placeholder: {
      control: 'text',
    },
    value: {
      control: 'text',
    },
    defaultValue: {
      control: 'text',
    },
    onValueChange: {
      action: 'onValueChange',
    },
    isOpen: {
      control: 'boolean',
    },
    isDisabled: {
      control: 'boolean',
    },
    isDanger: {
      control: 'boolean',
    },
    onOpenChange: {
      action: 'onOpenChange',
    },
    onBlur: {
      action: 'onBlur',
    },
    triggerComponent: {
      control: 'node',
    },
    popoverProps: {
      control: 'object',
    },
  },
  args: {
    placeholder: 'Select an option',
    triggerComponent: <SelectButton>Select an option</SelectButton>,
  },
} as Meta<typeof SelectWrapper>;

const Template: StoryFn<typeof SelectWrapper> = (args) => {
  return (
    <SelectWrapper {...args}>
      <SelectOption value="Option 1">Option 1</SelectOption>
      <SelectOption value="Option 2">Option 2</SelectOption>
      <SelectOption value="Option 3">Option 3</SelectOption>
    </SelectWrapper>
  );
};

export const DefaultStory = Template.bind({});
DefaultStory.storyName = 'Default';

export const TriggerComponent = Template.bind({});
TriggerComponent.args = {
  placeholder: '',
  triggerComponent: <SelectButton>Trigger component</SelectButton>,
};

export const Children: StoryFn<typeof SelectWrapper> = (args) => {
  return (
    <>
      <SelectWrapper {...args}>
        <Text paddingLeft={2} paddingRight={2}>
          All elements contained in SelectWrapper will be rendered within the
          popover
        </Text>
        <SelectOption value="Child 1">Child 1</SelectOption>
        <SelectOption value="Child 2">Child 2</SelectOption>
        <SelectOption value="Child 3">Child 3</SelectOption>
      </SelectWrapper>
    </>
  );
};
Children.args = {
  placeholder: 'Children demo',
};

export const ControlledOpen: StoryFn<typeof SelectWrapper> = (args) => {
  const [isOpen, setIsOpen] = React.useState(false);

  return (
    <>
      <Text>
        This demo is using props <strong>isOpen</strong> and{' '}
        <strong>onOpenChange</strong>
      </Text>
      <SelectWrapper
        {...args}
        isOpen={isOpen}
        onOpenChange={setIsOpen}
        triggerComponent={
          <SelectButton onClick={() => setIsOpen(!isOpen)}>
            Select an option
          </SelectButton>
        }
      >
        <SelectOption value="Option 1">Option 1</SelectOption>
        <SelectOption value="Option 2">Option 2</SelectOption>
        <SelectOption value="Option 3">Option 3</SelectOption>
      </SelectWrapper>
    </>
  );
};
ControlledOpen.args = {
  placeholder: 'Controlled open demo',
};

export const UncontrolledOpen = Template.bind({});
UncontrolledOpen.args = {
  placeholder: 'Uncontrolled open demo',
};

export const ControlledValue: StoryFn<typeof SelectWrapper> = (args) => {
  const [controlledValue, setControlledValue] = React.useState<string>('');

  return (
    <>
      <Text>
        This demo is using props <strong>defaultValue</strong>,{' '}
        <strong>value</strong>, and <strong>onValueChange</strong>
      </Text>
      <SelectWrapper
        {...args}
        defaultValue={'DefaultValue'}
        value={controlledValue}
        onValueChange={(value) => setControlledValue(value)}
      >
        <SelectOption value="Option 1">Option 1</SelectOption>
        <SelectOption value="Option 2">Option 2</SelectOption>
        <SelectOption value="Option 3">Option 3</SelectOption>
      </SelectWrapper>
    </>
  );
};

export const UncontrolledValue = Template.bind({});
UncontrolledValue.args = {
  placeholder: 'Uncontrolled value demo',
};

export const UseSelectContext: StoryFn<typeof SelectWrapper> = (args) => {
  // Note that the SelectContext is being used inside a component, because the SelectContext needs to be called within the SelectWrapper component and not before
  const CustomClose = () => {
    const { toggleUncontrolledOpen } = useSelectContext();

    return (
      <>
        <Text>Custom close button using SelectContext</Text>
        <Button block onClick={toggleUncontrolledOpen}>
          Close
        </Button>
      </>
    );
  };

  return (
    <>
      <SelectWrapper {...args}>
        <CustomClose />
        <SelectOption value="Option 1">Option 1</SelectOption>
        <SelectOption value="Option 2">Option 2</SelectOption>
        <SelectOption value="Option 3">Option 3</SelectOption>
      </SelectWrapper>
    </>
  );
};
UseSelectContext.args = {
  placeholder: 'useSelectContext demo',
};
UseSelectContext.storyName = 'useSelectContext';

export const IsMultiSelect: StoryFn<typeof SelectWrapper> = (args) => {
  return (
    <>
      <Text>
        When <strong>isMultiSelect</strong> prop is <i>true</i> the popover will
        remain open after a selection is made
      </Text>
      <SelectWrapper
        {...args}
        triggerComponent={<SelectButton>Demo</SelectButton>}
      >
        <SelectOption value="Option 1">Option 1</SelectOption>
        <SelectOption value="Option 2">Option 2</SelectOption>
        <SelectOption value="Option 3">Option 3</SelectOption>
      </SelectWrapper>
    </>
  );
};
IsMultiSelect.args = {
  placeholder: 'isMultiSelect demo',
  isMultiSelect: true,
};
IsMultiSelect.storyName = 'isMultiSelect';

export const Placeholder = Template.bind({});
Placeholder.args = {
  placeholder: 'Placeholder demo',
};

export const IsDisabled = Template.bind({});
IsDisabled.args = {
  placeholder: 'isDisabled demo',
  isDisabled: true,
};
IsDisabled.storyName = 'isDisabled';

export const IsDanger = Template.bind({});
IsDanger.args = {
  placeholder: 'isDanger demo',
  isDanger: true,
};
IsDanger.storyName = 'isDanger';

export const PopoverProps = Template.bind({});
PopoverProps.args = {
  placeholder: 'popoverProps demo',
  popoverProps: {
    backgroundColor: BackgroundColor.goerli,
    padding: 4,
    isPortal: false,
  },
};
PopoverProps.storyName = 'popoverProps';

export const OnBlur: StoryFn<typeof SelectWrapper> = (args) => {
  const [onBlur, setOnBlur] = React.useState(0);
  return (
    <>
      <SelectWrapper
        {...args}
        onBlur={() => setOnBlur(onBlur + 1)}
        triggerComponent={<SelectButton>onBlur count: {onBlur}</SelectButton>}
      >
        <Text>This is a demo of controlled onBlur.</Text>
      </SelectWrapper>
    </>
  );
};
OnBlur.args = {
  placeholder: '',
};
OnBlur.storyName = 'onBlur';
