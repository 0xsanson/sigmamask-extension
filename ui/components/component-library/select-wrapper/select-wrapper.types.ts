import type {
  StyleUtilityProps,
  PolymorphicComponentPropWithRef,
} from '../box';
import { PopoverStyleUtilityProps } from '../popover';

export type SelectContextType = {
  isOpen: boolean | undefined;
  onOpenChange: React.Dispatch<React.SetStateAction<boolean | undefined>>; // Double check this is correct type
  isUncontrolledOpen: boolean;
  setIsUncontrolledOpen: React.Dispatch<React.SetStateAction<any | null>>;
  toggleUncontrolledOpen: () => void; // Function to quickly toggle the open state
  isDisabled: boolean;
  isMultiSelect: boolean;
  value: any | null;
  onValueChange?: any;
  uncontrolledValue: any | null;
  setUncontrolledValue: React.Dispatch<React.SetStateAction<any | null>>;
  defaultValue: any | null;
  placeholder: any;
  isDanger: boolean;
  // onBlur?: React.FocusEventHandler;
  // onFocus?: React.FocusEventHandler;
};

export interface SelectWrapperStyleUtilityProps extends StyleUtilityProps {
  /*
   * Additional classNames to be added to the SelectWrapper component
   */
  className?: string;
  /*
   * Children of SelectWrapper component are put inside the select menu which uses the Popover component
   */
  children?: React.ReactNode;
  /*
   * Placeholder for SelectWrapper component to be displayed when no value or defaultValue is set
   * Can be accessed within any component inside the SelectWrapper
   */
  placeholder?: any;
  /*
   * Selected value of SelectWrapper component
   * Can be accessed within any component inside the SelectWrapper
   */
  value?: any;
  /**
   * Default value of SelectWrapper component
   * Can be accessed within any component inside the SelectWrapper
   */
  defaultValue?: any;
  /*
   * TODO: write info about type here
   */
  onValueChange?: any;
  /*
   * TODO: write info about type here
   */
  name?: string;
  /**
   * TODO: write info about type here
   */
  isOpen?: boolean;
  /*
   * TODO: write info about type here
   */
  onFocus?: () => void;
  /*
   * TODO: write info about type here
   */
  onBlur?: () => void;
  /*
   * TODO: write info about type here
   */
  triggerComponent: any;
  /*
   * TODO: write info about type here
   */
  popoverProps?: PopoverStyleUtilityProps;
}

export type SelectWrapperProps<C extends React.ElementType> =
  PolymorphicComponentPropWithRef<C, SelectWrapperStyleUtilityProps>;

export type SelectWrapperComponent = <C extends React.ElementType = 'div'>(
  props: SelectWrapperProps<C>,
) => React.ReactElement | null;
