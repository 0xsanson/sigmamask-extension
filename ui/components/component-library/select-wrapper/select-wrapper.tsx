import React, { useState } from 'react';
import classnames from 'classnames';
import type { PolymorphicRef, BoxProps } from '../box';
import { Box, Popover, PopoverPosition } from '..';
import {
  SelectWrapperComponent,
  SelectWrapperProps,
} from './select-wrapper.types';
import { SelectContext } from './select-wrapper.context';

export const SelectWrapper: SelectWrapperComponent = React.forwardRef(
  <C extends React.ElementType = 'div'>(
    {
      className = '',
      defaultValue,
      value,
      onValueChange,
      placeholder,
      isDanger,
      isDisabled,
      isOpen,
      onOpenChange,
      isMultiSelect, // Prevents the uncontrolled open state from being toggled
      triggerComponent,
      popoverProps,
      children,
      onBlur, // Controlled onBlur prop
      ...props
    }: SelectWrapperProps<C>,
    ref?: PolymorphicRef<C>,
  ) => {
    const [uncontrolledValue, setUncontrolledValue] = useState<any | null>();
    const [isUncontrolledOpen, setIsUncontrolledOpen] =
      useState<boolean>(false);
    const [referenceElement, setReferenceElement] =
      useState<HTMLElement | null>();

    const setBoxRef = (anchorRef: HTMLElement | null) => {
      setReferenceElement(anchorRef);
    };

    const toggleUncontrolledOpen = () => {
      setIsUncontrolledOpen(!isUncontrolledOpen);
    };

    const handleBlur = (e: any) => {
      const wrapper = e.currentTarget;
      const { relatedTarget } = e;

      if (!wrapper.contains(relatedTarget)) {
        // Close the popover only if the related target is not inside the wrapper
        setIsUncontrolledOpen(false);
        // If you have a controlled isOpen state, update it to close
        if (onOpenChange) {
          onOpenChange(false);
        }
        // Allow the dev to pass in a controlled onBlur prop
        if (onBlur) {
          onBlur(e);
        }
      }
    };

    return (
      <SelectContext.Provider
        value={{
          isOpen,
          onOpenChange,
          isUncontrolledOpen,
          setIsUncontrolledOpen,
          toggleUncontrolledOpen,
          isDisabled,
          isDanger,
          defaultValue,
          value,
          onValueChange,
          uncontrolledValue,
          setUncontrolledValue,
          placeholder,
          isMultiSelect,
        }}
      >
        <Box
          className={classnames('mm-select-wrapper', className)}
          onBlur={handleBlur} // This only works if the triggerComponent is a focusable element like a button so the onBlur event bubbles up
          ref={ref}
          {...(props as BoxProps<C>)}
        >
          {triggerComponent &&
            React.cloneElement(triggerComponent, {
              ref: setBoxRef,
            })}
          <Popover
            isOpen={isOpen || isUncontrolledOpen}
            position={PopoverPosition.Bottom}
            matchWidth
            referenceElement={referenceElement}
            padding={0}
            {...popoverProps}
            className={classnames(
              'mm-select-wrapper__popover',
              popoverProps?.className || '',
            )}
          >
            {children}
          </Popover>
        </Box>
      </SelectContext.Provider>
    );
  },
);
