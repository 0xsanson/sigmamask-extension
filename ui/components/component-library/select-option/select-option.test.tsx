import * as React from 'react';
import { render } from '@testing-library/react';
import { SelectWrapper } from '../select-wrapper';
import { SelectOption } from '.';

describe('SelectOption', () => {
  it('should render the SelectOption without crashing', () => {
    const { container } = render(
      <SelectWrapper isOpen={true} triggerComponent={<button>Trigger</button>}>
        <SelectOption />
      </SelectWrapper>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render the SelectOption with additional className', () => {
    const { getByTestId } = render(
      <SelectWrapper isOpen={true} triggerComponent={<button>Trigger</button>}>
        <SelectOption data-testid="classname" className="mm-test" />
      </SelectWrapper>,
    );
    expect(getByTestId('classname')).toHaveClass('mm-select-option mm-test');
  });
});
