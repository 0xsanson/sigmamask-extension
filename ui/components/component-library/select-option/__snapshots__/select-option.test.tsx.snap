// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectOption should render the SelectOption without crashing 1`] = `
<div>
  <div
    class="mm-box mm-select-wrapper"
  >
    <button>
      Trigger
    </button>
    <div
      class="mm-box mm-popover mm-popover--open mm-popover--reference-hidden mm-select-wrapper__popover mm-box--padding-0 mm-box--background-color-background-default mm-box--rounded-lg mm-box--border-color-border-muted box--border-style-solid box--border-width-1"
      role="tooltip"
      style="position: absolute; left: 0px; top: 0px; width: 0px;"
    >
      <button
        class="mm-box mm-select-option mm-box--display-block"
      />
    </div>
  </div>
</div>
`;
